extensions/grpc/codegen/src/main/java/io/quarkus/grpc/deployment/GrpcCodeGen.java

```diff
@@ -262,6 +270,25 @@ private boolean isGeneratingFromAppDependenciesEnabled(Config config) {
                 .filter(value -> !"none".equals(value)).isPresent();
     }
 
+    private boolean shouldGenerateDescriptorSet(Config config) {
+        return config.getOptionalValue(GENERATE_DESCRIPTOR_SET, Boolean.class).orElse(FALSE);
+    }
+
+    private Path getDescriptorSetOutputFile(CodeGenContext context) throws IOException {
+        var dscOutputDir = context.config().getOptionalValue(DESCRIPTOR_SET_OUTPUT_DIR, String.class)
+                .map(context.projectDir()::resolve)
```

Comment:
Is this the only need for projectDir?
I guess it simplifies the desc out-dir usage?
Or it could be still done with full relative path?

Reply:
> Or it could be still done with full relative path?

I'm not sure what you mean by this?

We need a reference to the project directory because the user could specify a relative path to it for the descriptor set output. Something like `src/test/resources` could be a good place to have it generated (so that tests could see it). Tools like Wiremock require configuring a path relative to the project root for where to pick it up (see https://wiremock.org/docs/grpc).

I didn't want to make an assumption that `$context.outDir/..` was _ALWAYS_ the project root dir, because that might not always be true. I thought it was safer to have the build tool pass it in.

Reply:
So ... is projectDir the root of the overall project (even in multi-module apps) or just a directory of your current module (sub-app if you call it)?
e.g. if I have a project with multiple microservices apps

Reply:
Its the directory of your current module. The current application being built.

Reply:
Just wondering "out-loud", since I'm surprised we didn't need this dir so far.

Reply:
Where does then `$context.outDir/..` point to, if not to the module's root?

Reply:
But yeah, I see your point.
It's good to have project-dir there, specially being 100% sure knowing where it points to.
And hence you can then have things like you described - placing things easily into `src/test/recources`, etc

Reply:
I thought this as well and tried to play around with just using `outDir()` (essentially the project's build dir), but then I kept thinking that it wouldn't be an unreasonable requirement to have some process generate these files and then commit to source control somewhere. Or do you think that requirement is unreasonable? Honestly I could argue it either way. I just thought this was more flexible in the end.

Supposedly there is `CodeGenContext.applicationModel().getApplicationModule()` which has `getModuleDir()`, which would provide this info, but during my testing I found that `CodeGenContext.applicationModel().getApplicationModule()` was always returning `null`.

This code is pretty unfamiliar to me and I couldn't figure out how/where it was getting constructed, so I opted for what I ended up doing.

Reply:
> Where does then `$context.outDir/..` point to, if not to the module's root?

From the docs: `context.outDir/..`would typically be resolved as `<project.build.directory>/generated-sources/<codegen-provider-id>`

Reply:
> but then I kept thinking that it wouldn't be an unreasonable requirement to have some process generate these files and then commit to source control somewhere. Or do you think that requirement is unreasonable? Honestly I could argue it either way. I just thought this was more flexible in the end.

Well, generating things directly into source dirs ... never really the best idea, but ... if it's needed it needed. :-)

> This code is pretty unfamiliar to me and I couldn't figure out how/where it was getting constructed, so I opted for what I ended up doing.

This is imho a question for @aloubyansky ...

Reply:
> Well, generating things directly into source dirs ... never really the best idea

Generally I would agree, but a descriptor set's intent is to be able to be used by code/tooling. Its a self-describing artifact of the entire contract.

Again, though, I could argue it either way, and generally when I'm in that situation I tend to make things as flexible as possible. But if people think it would be better to keep generated things in the build directory then I can re-work this a bit. It will certainly simplify the implementation.

Reply:
@cescoffier any thoughts?

Reply:
About the location of the generated descriptor, we must be sure the generation is idempotent, even when using different os/ machines...

Reply:
I guess by allowing it to be outside the build directory (`context.workDir`), then running a mvn/gradle `clean` will NOT remove it. Once its generated, its generated.

The more I think about it the more I think we should keep it confined to the `workDir` and not allow the generation outside of that. It is, after all, a generated resource and should be treated as such.

If you both (@cescoffier / @alesj ) are ok with that, I'll re-work this PR for that. It'll make it much simpler actually.

Reply:
> Supposedly there is `CodeGenContext.applicationModel().getApplicationModule()` which has `getModuleDir()`, which would provide this info, but during my testing I found that `CodeGenContext.applicationModel().getApplicationModule()` was always returning `null`.
> 
> This code is pretty unfamiliar to me and I couldn't figure out how/where it was getting constructed, so I opted for what I ended up doing.

The module layout part of that API is, by default, initialized only in dev and test modes, since this is when we need to load classes and resources from the workspace instead of from the JARs.
However, it can be enabled for the package building mode too by enabling a project property. Perhaps it should always be available. There were other use-cases where this was necessary. The reason not to do that was to not waste time discovering the workspace.

Reply:
Thank you for the info!

Reply:
> It is, after all, a generated resource and should be treated as such.

Exactly.
All other generated stuff goes there; 
e.g. proto sources, etc ... and tools (mostly) don't have problems with it.
Imho, this Wiremock should be flexible enough to handle this as well.

Reply:
I've made the change. I still have all the same integration tests, but now there is less code in the actual codebase that has changed. Just the `GrpcCodeGen` class.